test/lit/passes/unsubtyping.wast

```diff
@@ -0,0 +1,1599 @@
+ ;; CHECK:       (type $sub (sub $super (struct )))
+ (type $sub (sub $super (struct)))
+
+ ;; CHECK:       (type $2 (func))
+
+ ;; CHECK:       (type $3 (func (param (ref $super))))
+
+ ;; CHECK:      (type $4 (func (param (ref $super))))
+
+ ;; CHECK:      (tag $t (param (ref $super)))
+ (tag $t (param (ref $super)))
+
+ ;; CHECK:      (func $throw (type $2)
+ ;; CHECK-NEXT:  (throw $t
+ ;; CHECK-NEXT:   (struct.new_default $sub)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $throw
+  (throw $t
+   ;; This requires $sub <: $super.
+   (struct.new $sub)
+  )
+ )
+)
+
+(module
+ ;; CHECK:      (rec
+ ;; CHECK-NEXT:  (type $super (sub (struct )))
+ (type $super (sub (struct)))
+ ;; CHECK:       (type $sub (sub $super (struct )))
+ (type $sub (sub $super (struct)))
+
+ ;; CHECK:       (type $f (func (param (ref $super))))
+ (type $f (func (param (ref $super))))
+
+ ;; CHECK:      (elem declare func $call-ref)
+
+ ;; CHECK:      (func $call-ref (type $f) (param $0 (ref $super))
+ ;; CHECK-NEXT:  (call_ref $f
+ ;; CHECK-NEXT:   (struct.new_default $sub)
+ ;; CHECK-NEXT:   (ref.func $call-ref)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (block ;; (replaces something unreachable we can't emit)
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (struct.new_default $sub)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (unreachable)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (unreachable)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (block ;; (replaces something unreachable we can't emit)
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (struct.new_default $sub)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (ref.null nofunc)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (unreachable)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $call-ref (type $f) (param (ref $super))
+  ;; This requires $sub <: $super.
+  (call_ref $f
+   (struct.new $sub)
+   (ref.func $call-ref)
+  )
+  ;; This should not trip us up.
+  (call_ref $f
+   (struct.new $sub)
+   (unreachable)
+  )
+  ;; Nor should this.
+  (call_ref $f
+   (struct.new $sub)
+   (ref.null nofunc)
+  )
+ )
+)
+
+(module
+ ;; CHECK:      (rec
+ ;; CHECK-NEXT:  (type $super (sub (struct )))
+ (type $super (sub (struct)))
+ ;; CHECK:       (type $sub (sub $super (struct )))
+ (type $sub (sub $super (struct)))
+
+ ;; CHECK:       (type $f (func (result (ref $sub))))
+ (type $f (func (result (ref $sub))))
+
+ ;; CHECK:       (type $3 (func (result (ref $super))))
+
+ ;; CHECK:      (elem declare func $callee)
+
+ ;; CHECK:      (func $return-call-ref (type $3) (result (ref $super))
+ ;; CHECK-NEXT:  (return_call_ref $f
+ ;; CHECK-NEXT:   (ref.func $callee)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $return-call-ref (result (ref $super))
+  ;; This requires $sub <: $super.
+  (return_call_ref $f
+   (ref.func $callee)
+  )
+ )
+
+ ;; CHECK:      (func $callee (type $f) (result (ref $sub))
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $callee (result (ref $sub))
+  (unreachable)
+ )
+)
+
```

Comment:
IIANM, we are missing a test for the `call_ref` target's heap type itself (not its params and results)? I mean:
```wat
(call_ref $super
  (ref.func $sub_func)
)

(func $sub_func (type $sub))

(type $sub (sub $super (func)))
```

Reply:
Regardless of that being a call_ref, we are writing a thing of type `$sub` into a lot of type `$super` (on function types).

Reply:
Oh, wait, I forgot that in Binaryen IR we don't keep a separate (redundant) annotation. So the annotation is always equal to the type of the given target anyhow.